client/executor/wasmtime/src/runtime.rs

```diff
@@ -373,18 +414,47 @@ pub struct DeterministicStackLimit {
 	pub native_stack_max: u32,
 }
 
-pub struct Semantics {
-	/// Enabling this will lead to some optimization shenanigans that make calling [`WasmInstance`]
-	/// extremely fast.
-	///
-	/// Primarily this is achieved by not recreating the instance for each call and performing a
-	/// bare minimum clean up: reapplying the data segments and restoring the values for global
-	/// variables.
+/// The instantiation strategy to use for the WASM executor.
+///
+/// All of the CoW strategies (with `CopyOnWrite` suffix) are only supported when:
+///   a) we're running on Linux,
+///   b) we're running on an Unix-like system and we're precompiling
+///      our module beforehand.
+///
+/// If the CoW variant of a strategy is unsupported the executor will
+/// fall back to the non-CoW equivalent.
+#[non_exhaustive]
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub enum InstantiationStrategy {
+	/// Pool the instances to avoid initializing everything from scratch
+	/// on each instantiation. Use copy-on-write memory when possible.
 	///
-	/// Since this feature depends on instrumentation, it can be set only if runtime is
-	/// instantiated using the runtime blob, e.g. using [`create_runtime`].
-	// I.e. if [`CodeSupplyMode::Verbatim`] is used.
-	pub fast_instance_reuse: bool,
+	/// This is the fastest instantiation strategy.
+	PoolingCopyOnWrite,
+
+	/// Recreate the instance from scratch on every instantiation.
+	/// Use copy-on-write memory when possible.
+	RecreateInstanceCopyOnWrite,
+
+	/// Pool the instances to avoid initializing everything from scratch
+	/// on each instantiation.
+	Pooling,
+
+	/// Recreate the instance from scratch on every instantiation. Very slow.
+	RecreateInstance,
+
+	/// Legacy instance reuse mechanism. DEPRECATED. Will be removed. Do not use.
+	LegacyInstanceReuse,
+}
+
+enum InstantiationStrategyWithData {
+	LegacyInstanceReuse(InstanceSnapshotData),
+	Native,
```

Comment:
Nit: The term "Native" is a bit confusing, since there is already this term in scope as in `native_stack_max`.

 Can `Builtin` be better? Maybe `WasmtimeBuiltin`?

We might consider changing the name as well, since these looks like optional instantiation parameters. They are passed with our homebrew version of reusing is used, but they do not need to be passed in the case when the builtin mechanism is used.

Reply:
This is only an internal enum which we can remove once `legacy-instance-reuse` is gone, so I don't think it matters *that* much, but sounds good to me; I'll rename it.

Reply:
Sure. In general if I don't have a strong opinion about something (and I usually don't) I'm happy to oblige. (: